File: Pletor-canvas/backend/src/services/agentExecutionService.ts
import Anthropic from '@anthropic-ai/sdk'
import { prisma } from '../config/database.js'
import { env } from '../config/env.js'
import { AppError } from '../middlewares/errorHandler.js'
import { buildAgentContext, buildPromptFromContext } from './agentContextBuilder.js'

const SYSTEM_PROMPT = `Jsi AI agent v systému Pletor — vizuálním orchestračním IDE.
Generuješ kód na základě kontextu z WorkFlowy a canvas závislostí.
Piš čistý, produkční TypeScript kód. Dodržuj Clean Architecture.
Nepiš komentáře ke zjevným věcem. Používej early returns.
Odpovídej POUZE kódem — žádné vysvětlování, žádný markdown.`

function getClient(): Anthropic {
  if (!env.ANTHROPIC_API_KEY) {
    throw new AppError(400, 'ANTHROPIC_NOT_CONFIGURED', 'ANTHROPIC_API_KEY není nastaven v .env')
  }
  return new Anthropic({ apiKey: env.ANTHROPIC_API_KEY })
}

// Spustí generování kódu pro canvas uzel — non-streaming (jednoduchá verze)
export async function executeAgent(canvasNodeId: string, model?: string) {
  const agentContext = await buildAgentContext(canvasNodeId)
  const prompt = buildPromptFromContext(agentContext)
  const selectedModel = model ?? 'claude-sonnet-4-20250514'

  // Vytvoř záznam o spuštění
  const execution = await prisma.agentExecution.create({
    data: {
      canvasNodeId,
      prompt,
      context: JSON.stringify(agentContext),
      model: selectedModel,
      status: 'running',
    },
  })

  const startTime = Date.now()

  const client = getClient()

  const response = await client.messages.create({
    model: selectedModel,
    max_tokens: 4096,
    system: SYSTEM_PROMPT,
    messages: [{ role: 'user', content: prompt }],
  })

  const output = response.content
    .filter((block) => block.type === 'text')
    .map((block) => block.text)
    .join('\n')

  const durationMs = Date.now() - startTime

  // Aktualizuj záznam
  return prisma.agentExecution.update({
    where: { id: execution.id },
    data: {
      status: 'completed',
      output,
      tokensIn: response.usage.input_tokens,
      tokensOut: response.usage.output_tokens,
      durationMs,
    },
  })
}

// Streaming verze — vrátí AsyncIterable s chunky textu
export async function* executeAgentStream(canvasNodeId: string, model?: string) {
  const agentContext = await buildAgentContext(canvasNodeId)
  const prompt = buildPromptFromContext(agentContext)
  const selectedModel = model ?? 'claude-sonnet-4-20250514'

  const execution = await prisma.agentExecution.create({
    data: {
      canvasNodeId,
      prompt,
      context: JSON.stringify(agentContext),
      model: selectedModel,
      status: 'running',
    },
  })

  const startTime = Date.now()
  const client = getClient()

  const stream = client.messages.stream({
    model: selectedModel,
    max_tokens: 4096,
    system: SYSTEM_PROMPT,
    messages: [{ role: 'user', content: prompt }],
  })

  // Posílej chunky textu průběžně
  let fullOutput = ''

  // Pošli metadata o spuštění jako první událost
  yield JSON.stringify({ type: 'start', executionId: execution.id, model: selectedModel }) + '\n'

  for await (const event of stream) {
    if (event.type === 'content_block_delta' && event.delta.type === 'text_delta') {
      fullOutput += event.delta.text
      yield JSON.stringify({ type: 'delta', text: event.delta.text }) + '\n'
    }
  }

  const finalMessage = await stream.finalMessage()
  const durationMs = Date.now() - startTime

  // Ulož výsledek
  await prisma.agentExecution.update({
    where: { id: execution.id },
    data: {
      status: 'completed',
      output: fullOutput,
      tokensIn: finalMessage.usage.input_tokens,
      tokensOut: finalMessage.usage.output_tokens,
      durationMs,
    },
  })

  yield JSON.stringify({
    type: 'done',
    executionId: execution.id,
    tokensIn: finalMessage.usage.input_tokens,
    tokensOut: finalMessage.usage.output_tokens,
    durationMs,
  }) + '\n'
}

// Načte historii spuštění pro uzel
export function getExecutionsByNodeId(canvasNodeId: string) {
  return prisma.agentExecution.findMany({
    where: { canvasNodeId },
    orderBy: { createdAt: 'desc' },
    take: 10,
  })
}

// Načte konkrétní spuštění
export function getExecutionById(id: string) {
  return prisma.agentExecution.findUnique({ where: { id } })
}

File: Pletor-canvas/backend/app/agent/agent_execution_service.py
import json
import time
from collections.abc import AsyncGenerator
from typing import Any

import anthropic
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config.settings import settings
from app.middlewares.error_handler import AppError
from app.agent.agent_model import AgentExecution
from app.agent.agent_context_builder import build_agent_context, build_prompt_from_context

SYSTEM_PROMPT = """Jsi AI agent v systému Pletor — vizuálním orchestračním IDE.
Generuješ kód na základě kontextu z WorkFlowy a canvas závislostí.
Piš čistý, produkční TypeScript kód. Dodržuj Clean Architecture.
Nepiš komentáře ke zjevným věcem. Používej early returns.
Odpovídej POUZE kódem — žádné vysvětlování, žádný markdown."""


def _get_client() -> anthropic.AsyncAnthropic:
    if not settings.ANTHROPIC_API_KEY:
        raise AppError(400, "ANTHROPIC_NOT_CONFIGURED", "ANTHROPIC_API_KEY není nastaven v .env")
    return anthropic.AsyncAnthropic(api_key=settings.ANTHROPIC_API_KEY)


async def execute_agent(db: AsyncSession, canvas_node_id: str, model: str | None = None) -> AgentExecution:
    """Spustí generování kódu pro canvas uzel — non-streaming."""
    agent_context = await build_agent_context(db, canvas_node_id)
    prompt = build_prompt_from_context(agent_context)
    selected_model = model or "claude-sonnet-4-20250514"

    execution = AgentExecution(
        canvas_node_id=canvas_node_id,
        prompt=prompt,
        context=json.dumps(agent_context),
        model=selected_model,
        status="running",
    )
    db.add(execution)
    await db.commit()
    await db.refresh(execution)

    start_time = time.monotonic()
    client = _get_client()

    response = await client.messages.create(
        model=selected_model,
        max_tokens=4096,
        system=SYSTEM_PROMPT,
        messages=[{"role": "user", "content": prompt}],
    )

    output = "\n".join(block.text for block in response.content if block.type == "text")
    duration_ms = int((time.monotonic() - start_time) * 1000)

    execution.status = "completed"
    execution.output = output
    execution.tokens_in = response.usage.input_tokens
    execution.tokens_out = response.usage.output_tokens
    execution.duration_ms = duration_ms

    await db.commit()
    await db.refresh(execution)
    return execution


async def execute_agent_stream(db: AsyncSession, canvas_node_id: str, model: str | None = None) -> AsyncGenerator[str, None]:
    """Streaming verze — vrátí AsyncGenerator s chunky textu ve formátu SSE."""
    agent_context = await build_agent_context(db, canvas_node_id)
    prompt = build_prompt_from_context(agent_context)
    selected_model = model or "claude-sonnet-4-20250514"

    execution = AgentExecution(
        canvas_node_id=canvas_node_id,
        prompt=prompt,
        context=json.dumps(agent_context),
        model=selected_model,
        status="running",
    )
    db.add(execution)
    await db.commit()
    await db.refresh(execution)

    start_time = time.monotonic()
    client = _get_client()

    yield json.dumps({"type": "start", "executionId": execution.id, "model": selected_model}) + "\n"

    full_output = ""

    async with client.messages.stream(
        model=selected_model,
        max_tokens=4096,
        system=SYSTEM_PROMPT,
        messages=[{"role": "user", "content": prompt}],
    ) as stream:
        async for text in stream.text_stream:
            full_output += text
            yield json.dumps({"type": "delta", "text": text}) + "\n"

        final_message = await stream.get_final_message()

    duration_ms = int((time.monotonic() - start_time) * 1000)

    execution.status = "completed"
    execution.output = full_output
    execution.tokens_in = final_message.usage.input_tokens
    execution.tokens_out = final_message.usage.output_tokens
    execution.duration_ms = duration_ms

    await db.commit()

    yield json.dumps({
        "type": "done",
        "executionId": execution.id,
        "tokensIn": final_message.usage.input_tokens,
        "tokensOut": final_message.usage.output_tokens,
        "durationMs": duration_ms,
    }) + "\n"


async def get_executions_by_node_id(db: AsyncSession, canvas_node_id: str) -> list[AgentExecution]:
    stmt = select(AgentExecution).where(AgentExecution.canvas_node_id == canvas_node_id).order_by(AgentExecution.created_at.desc()).limit(10)
    result = await db.execute(stmt)
    return list(result.scalars().all())


async def get_execution_by_id(db: AsyncSession, execution_id: str) -> AgentExecution | None:
    return await db.get(AgentExecution, execution_id)
